account-compression/programs/spl-compression/src/lib.rs

```diff
@@ -0,0 +1,550 @@
+    pub fn replace_leaf(
+        ctx: Context<Modify>,
+        root: [u8; 32],
+        previous_leaf: [u8; 32],
+        new_leaf: [u8; 32],
+        index: u32,
+    ) -> Result<()> {
+        let mut merkle_roll_bytes = ctx.accounts.merkle_roll.try_borrow_mut_data()?;
+        let (header_bytes, rest) =
+            merkle_roll_bytes.split_at_mut(size_of::<ConcurrentMerkleTreeHeader>());
+
+        let header = Box::new(ConcurrentMerkleTreeHeader::try_from_slice(header_bytes)?);
+        assert_eq!(header.authority, ctx.accounts.authority.key());
+        let merkle_roll_size = merkle_roll_get_size!(header)?;
+        let (roll_bytes, canopy_bytes) = rest.split_at_mut(merkle_roll_size);
+
+        let mut proof = vec![];
+        for node in ctx.remaining_accounts.iter() {
+            proof.push(node.key().to_bytes());
+        }
+        fill_in_proof_from_canopy(canopy_bytes, header.max_depth, index, &mut proof)?;
+        let id = ctx.accounts.merkle_roll.key();
+        // A call is made to ConcurrentMerkleTree::set_leaf(root, previous_leaf, new_leaf, proof, index)
+        let change_log = merkle_roll_apply_fn!(
+            header,
+            id,
+            roll_bytes,
+            set_leaf,
+            root,
+            previous_leaf,
+            new_leaf,
+            &proof,
+            index,
+        )?;
+        wrap_event(change_log.try_to_vec()?, &ctx.accounts.candy_wrapper)?;
+        emit!(*change_log);
+        update_canopy(canopy_bytes, header.max_depth, Some(change_log))
+    }
+
+    /// Transfers `authority`.
+    /// Requires `authority` to sign
+    pub fn transfer_authority(
+        ctx: Context<TransferAuthority>,
+        new_authority: Pubkey,
+    ) -> Result<()> {
+        let mut merkle_roll_bytes = ctx.accounts.merkle_roll.try_borrow_mut_data()?;
+        let (mut header_bytes, _) =
+            merkle_roll_bytes.split_at_mut(size_of::<ConcurrentMerkleTreeHeader>());
+
+        let mut header = Box::new(ConcurrentMerkleTreeHeader::try_from_slice(header_bytes)?);
+        assert_eq!(header.authority, ctx.accounts.authority.key());
+
+        header.authority = new_authority;
+        msg!("Authority transferred to: {:?}", header.authority);
+        header.serialize(&mut header_bytes)?;
+
+        Ok(())
+    }
+
+    /// Verifies a provided proof and leaf.
+    /// If invalid, throws an error.
+    pub fn verify_leaf(
+        ctx: Context<VerifyLeaf>,
+        root: [u8; 32],
+        leaf: [u8; 32],
+        index: u32,
+    ) -> Result<()> {
+        let mut merkle_roll_bytes = ctx.accounts.merkle_roll.try_borrow_mut_data()?;
+        let (header_bytes, rest) =
+            merkle_roll_bytes.split_at_mut(size_of::<ConcurrentMerkleTreeHeader>());
+        let header = Box::new(ConcurrentMerkleTreeHeader::try_from_slice(header_bytes)?);
+        let merkle_roll_size = merkle_roll_get_size!(header)?;
+        let (roll_bytes, canopy_bytes) = rest.split_at_mut(merkle_roll_size);
```

Comment:
Up to you, but none of this is checked, so someone could spoof a merkle tree.  Probably not a big deal, but better to check to be sure

Reply:
Yeah might be good to add that the owner of the merkle_tree is the correct program. Practically doesn’t matter for any of the instructions that actually modify the data

Reply:
Added